module/zcommon/zfs_fletcher.c

```diff
@@ -222,30 +264,216 @@ fletcher_4_incremental_native(const void *buf, uint64_t size,
+		algo->fini(zcp);
+}
+
+void
+fletcher_4_byteswap(const void *buf, uint64_t size, zio_cksum_t *zcp)
+{
+	const fletcher_4_func_t *algo = fletcher_4_impl_get();
+
+	algo->init(zcp);
+	algo->compute_byteswap(buf, size, zcp);
+	if (algo->fini != NULL)
+		algo->fini(zcp);
+}
+
+void
+fletcher_4_incremental_native(const void *buf, uint64_t size,
+    zio_cksum_t *zcp)
+{
+	fletcher_4_scalar(buf, size, zcp);
+}
+
 void
 fletcher_4_incremental_byteswap(const void *buf, uint64_t size,
     zio_cksum_t *zcp)
 {
-	const uint32_t *ip = buf;
-	const uint32_t *ipend = ip + (size / sizeof (uint32_t));
-	uint64_t a, b, c, d;
+	fletcher_4_scalar_byteswap(buf, size, zcp);
+}
 
-	a = zcp->zc_word[0];
-	b = zcp->zc_word[1];
-	c = zcp->zc_word[2];
-	d = zcp->zc_word[3];
+void
+fletcher_4_init(void)
+{
+	const uint64_t const bench_ns = (50 * MICROSEC); /* 50ms */
+	unsigned long best_run_count = 0;
+	unsigned long best_run_index = 0;
+	const unsigned data_size = 4096;
+	char *databuf;
+	int i;
 
-	for (; ip < ipend; ip++) {
-		a += BSWAP_32(ip[0]);
-		b += a;
-		c += b;
-		d += c;
+	databuf = kmem_alloc(data_size, KM_SLEEP);
+	for (i = 0; i < ARRAY_SIZE(fletcher_4_algos); i++) {
+		const fletcher_4_func_t *algo = fletcher_4_algos[i];
+		kstat_named_t *stat = &fletcher_4_kstat_data[i];
+		unsigned long run_count = 0;
+		hrtime_t start;
+		zio_cksum_t zc;
+
+		strncpy(stat->name, algo->name, sizeof (stat->name) - 1);
+		stat->data_type = KSTAT_DATA_UINT64;
+		stat->value.ui64 = 0;
+
+		if (algo->valid != NULL && !algo->valid())
+			continue;
+
+		kpreempt_disable();
+		start = gethrtime();
+		algo->init(&zc);
+		do {
+			algo->compute(databuf, data_size, &zc);
+			run_count++;
+		} while (gethrtime() < start + bench_ns);
+		if (algo->fini != NULL)
+			algo->fini(&zc);
+		kpreempt_enable();
+
+		if (run_count > best_run_count) {
+			best_run_count = run_count;
+			best_run_index = i;
+		}
+
+		/*
+		 * Due to high overhead of gethrtime(), the performance data
+		 * here is inaccurate and much slower than it could be.
+		 * It's fine for our use though because only relative speed
+		 * is important.
+		 */
+		stat->value.ui64 = data_size * run_count *
+		    (NANOSEC / bench_ns) >> 20; /* by MB/s */
 	}
+	kmem_free(databuf, data_size);
 
-	ZIO_SET_CHECKSUM(zcp, a, b, c, d);
+	fletcher_4_fastest = fletcher_4_algos[best_run_index];
+
+	mutex_init(&fletcher_4_impl_selector_lock, NULL, MUTEX_DEFAULT, NULL);
+	fletcher_4_impl_set("fastest");
+
+	fletcher_4_kstat = kstat_create("zfs", 0, "fletcher_4_bench",
+	    "misc", KSTAT_TYPE_NAMED, ARRAY_SIZE(fletcher_4_algos),
+	    KSTAT_FLAG_VIRTUAL);
+	if (fletcher_4_kstat != NULL) {
+		fletcher_4_kstat->ks_data = fletcher_4_kstat_data;
+		kstat_install(fletcher_4_kstat);
+	}
+}
+
+void
+fletcher_4_fini(void)
+{
+	mutex_destroy(&fletcher_4_impl_selector_lock);
+	if (fletcher_4_kstat != NULL) {
+		kstat_delete(fletcher_4_kstat);
+		fletcher_4_kstat = NULL;
+	}
 }
 
 #if defined(_KERNEL) && defined(HAVE_SPL)
+
+static int
```

Comment:
I could reproduce the issue 1 & 2 on my local node. Can you please show me your test cases?
I did ```echo axv2 > /sys/module/zcommon/parameters/fletcher_4_impl_set ``` and it returned ```-bash: echo: write error: Invalid argument``` also I deliberately disabled avx2 and tried to write ```avx2``` it returned error as expected.

For the issue 3, since the other symbols in the same file are all started with ```fletcher_4```, do you think it would be better for me to change the module parameter to ```zfs_fletcher_4_impl_set``` but leave the symbol name as-is?